### PR TITLE
Issue/4437 Add serial number to the metadata of the payment intent

### LINKED ISSUE #4437: [Mobile Payments] Add the card reader serial number (ID) to the payment intent meta data
We are charged for each reader that has processed 3 or more authorizations in a calendar month. To help us validate that assessment, we should add the card reader serial number (e.g. "CBH204909005931") to the payment intent meta data.

This will allow our back end to identify which readers have processed authorizations for a given calendar month.

cc @ctarda @bmccotter @laurendavissmith

more context here: https://a8c.slack.com/archives/CGGCLBN58/p1626121182302500

### PR DESCRIPTION
**Pull request by kidinov**
Resolves #4437 

The PR adds a serial number to the metadata of the payment intent

### How to test

Make sure your store is eligible for in-person payments

* Connect to a reader (watch out for this known issue)
* Open detail of an unpaid order in USD
* Tap on the Collect Payment button
* Finish the payment
* Open Stripe's dashboard and search for the payment

image

Update release notes:

[x] I have considered if this change warrants user-facing release notes and have added them to RELEASE-NOTES.txt if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt [hunk 1]
```diff
@@ -146,7 +150,8 @@ class CardReaderPaymentViewModel
                 customerEmail = customerEmail.ifEmpty { null },
                 customerName = "${order.billingAddress.firstName} ${order.billingAddress.lastName}".ifBlank { null },
                 storeName = selectedSite.get().name.ifEmpty { null },
-                siteUrl = selectedSite.get().url.ifEmpty { null }
+                siteUrl = selectedSite.get().url.ifEmpty { null },
+                readerId = readerId ?: ""
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt]: ❓ I'm wondering if it'd make sense to retrieve the id internal within the CardReaderModule -> it feels a bit off that the client asks the module for an id and passes it back to the module right after.

File: libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/payments/MetaDataKeys.kt [hunk 2]
```diff
@@ -39,7 +39,12 @@ internal enum class MetaDataKeys(val key: String) {
      * See also PaymentIntent.PaymentTypes
      * This key is also used by the plugin when it creates payment intents.
      */
-    PAYMENT_TYPE("payment_type");
+    PAYMENT_TYPE("payment_type"),
+
+    /**
+     * Serial number of a reader which used to collect the payment
```
Comment (malinajirka) [inline libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/payments/MetaDataKeys.kt]: 🔍 Possible typo: "which **is** used". Wdyt?

File: libs/cardreader/src/test/java/com/woocommerce/android/cardreader/internal/payments/actions/CreatePaymentActionTest.kt [hunk 3]
```diff
@@ -203,6 +204,20 @@ internal class CreatePaymentActionTest {
         assertThat(captor.firstValue[MetaDataKeys.ORDER_ID.key]).isEqualTo(expected.toString())
     }
 
+    @Test
+    fun `when creating payment intent, then reader id set`() = runBlockingTest {
+        val readerId = "SM12345678"
+        whenever(terminal.createPaymentIntent(any(), any())).thenAnswer {
```
Comment (malinajirka) [inline libs/cardreader/src/test/java/com/woocommerce/android/cardreader/internal/payments/actions/CreatePaymentActionTest.kt]: 🔍 We don't need this `whenever(..)` with the latest changes from the parent branch.

**Review by malinajirka — COMMENTED** [review]
Looks good 👏 , I've left some minor questions/suggestions.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt (hunk 1 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt]: Agree, I had exactly the same thought and even implemented it firstly the way you are talking about. But it either means that we have to make `PaymentInfo` mutable, or we have to pass this `readerId` as a second parameter to the several methods. To be honest I still don't know what approach is better here. Both are kind of not good. What do you think?

File: libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/payments/MetaDataKeys.kt (hunk 2 above)
Comment (kidinov) [reply libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/payments/MetaDataKeys.kt]: Fixed! https://github.com/woocommerce/woocommerce-android/pull/4470/commits/9be6f7b8446ab1f0c41ce2c0c34d32aa6347b2d1

File: libs/cardreader/src/test/java/com/woocommerce/android/cardreader/internal/payments/actions/CreatePaymentActionTest.kt (hunk 3 above)
Comment (kidinov) [reply libs/cardreader/src/test/java/com/woocommerce/android/cardreader/internal/payments/actions/CreatePaymentActionTest.kt]: Fixed! https://github.com/woocommerce/woocommerce-android/pull/4470/commits/13e85ce658098bbf8125e9498c267356fddacacb

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt (hunk 1 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt]: We already inject `TerminalWrapper` into `CreatePaymentAction`, so I'd consider using `terminal.getConnectedReader?.id` directly. Wdyt?

Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt]: Done this way. Please check  https://github.com/woocommerce/woocommerce-android/pull/4470/commits/1e4c677e33845f0808b89e77e2c2426b89a56d7b and https://github.com/woocommerce/woocommerce-android/pull/4470/commits/0f3502dacd02a0c9ed8fa7768b882dec7eb97072

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt [hunk 4]
```diff
@@ -146,7 +146,7 @@ class CardReaderPaymentViewModel
                 customerEmail = customerEmail.ifEmpty { null },
                 customerName = "${order.billingAddress.firstName} ${order.billingAddress.lastName}".ifBlank { null },
                 storeName = selectedSite.get().name.ifEmpty { null },
-                siteUrl = selectedSite.get().url.ifEmpty { null }
+                siteUrl = selectedSite.get().url.ifEmpty { null },
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:149]: 🔍 There are a few places in this PR where commas were added to the end of parameters - is that intentional?

**Review by malinajirka — APPROVED** [review]
Thanks @kidinov, it looks good to me. I've left one super minor comment - detekt doesn't complain so feel free to merge this PR, I was just curious if the changes were intentional.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt (hunk 4 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:149]: Yes, it's actually in the convention - https://kotlinlang.org/docs/coding-conventions.html#trailing-commas

Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:149]: Ohh, nice - I haven't read the guidelines after the release of 1.4! TIL thanks

**Pull request merged**